File: java-common/src/main/java/com/guusto/service/CrudService.java
package com.guusto.service;

import java.io.Serializable;
import java.util.Collection;

import com.guusto.mapping.entity.GenericEntity;
import com.guusto.mapping.entity.GenericModel;
import com.guusto.mapping.specification.Specification;
import com.guusto.query.PreparedQuery;

public interface CrudService<Model extends GenericModel, Entity extends GenericEntity> {

	void create(Model dto);

	Model update(Model dto);

	Collection<Model> findAll(PreparedQuery preparedQuery);

	Collection<Model> findAllWithoutLimit();

	Model find(Specification specification);

	Model findById(Serializable id);

	boolean contains(Model dto);

	boolean contains(Specification specification);

	Long count(Specification specification);

	Long count(PreparedQuery preparedQuery);
	
}

File: angular-guusto/src/app/shared/service/crud/crud-service.ts
import { PreparedQuery } from '../../query/prepared-query/prepared-query';
import { Observable } from 'rxjs';

export interface CrudService<T> {

    persist(model: T): Observable<any>;

    update(id: number, model: T): Observable<T>;

    remove(model: T): Observable<any>;

    removeById(id: number): Observable<any>;

    find(preparedQuery: PreparedQuery): Observable<T>;

    findAll(preparedQuery: PreparedQuery): Observable<T[]>;

    search(preparedQuery: PreparedQuery): Observable<T[]>;

    count(preparedQuery: PreparedQuery): Observable<number>;

    contains(model: T): Observable<boolean>;

}
